libraries/networking/src/udt/CongestionControl.h

```diff
@@ -0,0 +1,125 @@
+
+class Connection;
+class Packet;
+
+class CongestionControl {
+    friend class Connection;
+public:
+
+    CongestionControl() {};
+    CongestionControl(int synInterval) : _synInterval(synInterval) {}
+    virtual ~CongestionControl() {}
+    
+    int synInterval() const { return _synInterval; }
+
+    virtual void init() {}
+    virtual void close() {}
+    virtual void onACK(SequenceNumber ackNum) {}
+    virtual void onLoss(SequenceNumber rangeStart, SequenceNumber rangeEnd) {}
+    
+protected:
+    void setAckInterval(int ackInterval) { _ackInterval = ackInterval; }
+    void setRto(int rto) { _userDefinedRto = true; _rto = rto; }
+    
+    void setMSS(int mss) { _mss = mss; }
+    void setMaxCongestionWindowSize(int window) { _maxCongestionWindowSize = window; }
+    void setBandwidth(int bandwidth) { _bandwidth = bandwidth; }
+    void setMaxBandwidth(int maxBandwidth) { _maxBandwidth = maxBandwidth; }
+    void setSendCurrentSequenceNumber(SequenceNumber seqNum) { _sendCurrSeqNum = seqNum; }
+    void setReceiveRate(int rate) { _receiveRate = rate; }
+    void setRTT(int rtt) { _rtt = rtt; }
+    void setPacketSendPeriod(double newSendPeriod); // call this internally to ensure send period doesn't go past max bandwidth
+    
+    double _packetSendPeriod { 1.0 }; // Packet sending period, in microseconds
+    double _congestionWindowSize { 16.0 }; // Congestion window size, in packets
+    
+    int _bandwidth { 0 }; // estimated bandwidth, packets per second
+    int _maxBandwidth { -1 }; // Maximum desired bandwidth, packets per second
+    double _maxCongestionWindowSize { 0.0 }; // maximum cwnd size, in packets
+    
+    int _mss { 0 }; // Maximum Packet Size, including all packet headers
+    SequenceNumber _sendCurrSeqNum; // current maximum seq num sent out
+    int _receiveRate { 0 }; // packet arrive rate at receiver side, packets per second
+    int _rtt { 0 }; // current estimated RTT, microsecond
+    
+private:
+    CongestionControl(const CongestionControl& other) = delete;
+    CongestionControl& operator=(const CongestionControl& other) = delete;
+    
+    int _ackInterval { 0 }; // How many packets to send one ACK, in packets
+    int _lightACKInterval { 64 }; // How many packets to send one light ACK, in packets
+    
+    int _synInterval { DEFAULT_SYN_INTERVAL };
+    
+    bool _userDefinedRto { false }; // if the RTO value is defined by users
+    int _rto { -1 }; // RTO value, microseconds
+};
+    
+    
+class CongestionControlVirtualFactory {
+public:
+    virtual ~CongestionControlVirtualFactory() {}
+    
+    static int synInterval() { return DEFAULT_SYN_INTERVAL; }
+    
+    virtual std::unique_ptr<CongestionControl> create() = 0;
+};
+
+template <class T> class CongestionControlFactory: public CongestionControlVirtualFactory
+{
+public:
+    virtual ~CongestionControlFactory() {}
+    virtual std::unique_ptr<CongestionControl> create() { return std::unique_ptr<T>(new T()); }
```

Comment:
It's only in C++14
On Thu, Aug 6, 2015 at 11:06 AM Ryan Huffman <notifications@github.com>
wrote:

> In libraries/networking/src/udt/CongestionControl.h
> <https://github.com/birarda/hifi/pull/12#discussion_r36445708>:
>
> > +
> > +
> > +class CongestionControlVirtualFactory {
> > +public:
> > +    virtual ~CongestionControlVirtualFactory() {}
> > +
> > +    static int synInterval() { return DEFAULT_SYN_INTERVAL; }
> > +
> > +    virtual std::unique_ptr<CongestionControl> create() = 0;
> > +};
> > +
> > +template <class T> class CongestionControlFactory: public CongestionControlVirtualFactory
> > +{
> > +public:
> > +    virtual ~CongestionControlFactory() {}
> > +    virtual std::unique_ptr<CongestionControl> create() { return std::unique_ptr<T>(new T()); }
>
> make_unique
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/birarda/hifi/pull/12/files#r36445708>.
>